### PR TITLE
Closes #888 - Adds support for additional HTTP request methods

### LINKED ISSUE #888: HttpRequest does not support all HTTP request methods
`HttpRequest` supports only a subset of HTTP methods. It needs to support all types.

Currently supported:
- GET
- POST

Missing:
- HEAD
- PUT
- DELETE
- TRACE
- OPTIONS
- CONNECT
- PATCH

Additional considerations:
- Telemetry behavior of GET/POST should apply to other methods

Normative Refs:
- [RFC-7231, section 4](https://tools.ietf.org/html/rfc7231#section-4)
- [RFC-5789, section 2](https://tools.ietf.org/html/rfc5789#section-2)

Tracking:
- [PBI-889784](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/889784)

### PR DESCRIPTION
**Pull request by iambmelt**
See #888

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 1]
```diff
@@ -158,27 +166,83 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+
+    public static HttpResponse sendOptions(@NonNull final URL requestUrl,
+                                           @NonNull final Map<String, String> requestHeaders) {
+        // TODO
+        return null;
+    }
+
+    public static HttpResponse sendConnect(@NonNull final URL requestUrl,
+                                           @NonNull final Map<String, String> requestHeaders) {
+        // TODO Can this be supported? I suspect not.
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:234]: I don't think we can support this currently -- support may also not be needed

**Comment by iambmelt** [comment]
Consider adding a method which accepts an HTTP method and all optional params, that way the caller doesn't need to `if/else` their way to the right method

**Comment by iambmelt** [comment]
> Consider adding a method which accepts an HTTP method and all optional params, that way the caller doesn't need to `if/else` their way to the right method

Added in e0966fbc2d63835466991d86d432d168378d5294

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 2]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
 
-        final HttpEndEvent httpEndEvent = new HttpEndEvent();
-        if (response != null) {
-            httpEndEvent.putStatusCode(response.getStatusCode());
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendHead(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:228]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 3]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendPut(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:250]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 4]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendDelete(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:276]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 5]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendTrace(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:302]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 6]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendOptions(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:324]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 7]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendPatch(@NonNull final URL requestUrl,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:346]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java [hunk 8]
```diff
@@ -158,24 +203,223 @@ public static HttpResponse sendPost(final URL requestUrl, final Map<String, Stri
...
+        );
+
+        final HttpResponse response = httpRequest.send();
+
+        recordHttpTelemetryEventEnd(response);
+
+        return response;
+    }
+
+    public static HttpResponse sendWithMethod(@NonNull final String httpMethod,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:376]: javadoc

Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:376]: Refactor other methods to call this one

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 2 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:214]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 3 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:236]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 4 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:260]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 5 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:282]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 6 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:302]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 7 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:324]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 8 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:350]: Resolved in 5bdd4b78e1e7fb084ef92f7264a10c6fc664480c

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java:350]: Resolved in f7703d44d3354e084c9cf46813ccd0b16b56cc5b

File: common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java (hunk 1 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/net/HttpRequest.java]: Not needed. Resolving

**Comment by iambmelt** [comment]
HTTP Layer tests are currently disabled, blocked:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/test/java/com/microsoft/identity/common/HttpRequestTest.java#L43-L48

Investigating....

**Comment by iambmelt** [comment]
> HTTP Layer tests are currently disabled, blocked:
> https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/test/java/com/microsoft/identity/common/HttpRequestTest.java#L43-L48
> 
> Investigating....

Unjamming in https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/896

File: common/src/androidTest/java/com/microsoft/identity/common/HttpRequestTest.java [hunk 9]
```diff
@@ -0,0 +1,816 @@
...
+import static com.microsoft.identity.common.internal.net.HttpUrlConnectionFactory.getMockedConnectionCountInQueue;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link HttpRequest}.
+ * <p>
+ * We need to fix setup of these tests as something is missing post moving from MSAL to common
```
Comment (shahzaibj) [inline common/src/androidTest/java/com/microsoft/identity/common/HttpRequestTest.java:56]: I think we can remove this comment now

Comment (iambmelt) [reply common/src/androidTest/java/com/microsoft/identity/common/HttpRequestTest.java:56]: Yup, good catch

Comment (iambmelt) [reply common/src/androidTest/java/com/microsoft/identity/common/HttpRequestTest.java]: Fixed in c17b8ccab1722b99bfea6ef691c12a6b92bdacc7

**Pull request merged**